File: jlpt/native_ios/Sources/Domain/Models/Study.swift
import Foundation

struct StudyItem<T>: Identifiable where T: Identifiable, T.ID == Int64 {
    let item: T
    var status: StudyStatus
    var id: Int64 { item.id }
}

struct ReviewResult {
    let status: StudyStatus
    let intervalDays: Int
    let nextReviewAt: Date?
}

struct ProgressSummary {
    let countsByItemTypeAndStatus: [StudyItemType: [StudyStatus: Int64]]

    func count(for itemType: StudyItemType, status: StudyStatus) -> Int64 {
        countsByItemTypeAndStatus[itemType]?[status] ?? 0
    }

    func total(for itemType: StudyItemType) -> Int64 {
        countsByItemTypeAndStatus[itemType]?.values.reduce(0, +) ?? 0
    }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/domain/model/Study.kt
package com.jlpt.android.domain.model

import java.time.Instant

data class StudyItem<T>(
    val item: T,
    val status: StudyStatus
)

data class ReviewResult(
    val status: StudyStatus,
    val intervalDays: Int,
    val nextReviewAt: Instant?
)

data class ProgressSummary(
    val countsByItemTypeAndStatus: Map<StudyItemType, Map<StudyStatus, Long>>
) {
    fun countFor(itemType: StudyItemType, status: StudyStatus): Long =
        countsByItemTypeAndStatus[itemType]?.get(status) ?: 0L

    fun totalFor(itemType: StudyItemType): Long =
        countsByItemTypeAndStatus[itemType]?.values?.sum() ?: 0L
}
